File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-1.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        left, right = 0, k + 1
        freq_counter = self.getFreqCounter(s[left : right])
        max_len = 0

        while right < len(s):
            freq_counter[ord(s[right]) - ord('A')] += 1

            while (right - left + 1) - max(freq_counter) > k:
                freq_counter[ord(s[left]) - ord('A')] -= 1
                left += 1
            max_len = max(max_len, right - left + 1)
            right += 1 
        
        return max_len
    
    def getFreqCounter(self, s: str) -> List[int]:
        freq_counter = [0] * 26
        for char in s:
            freq_counter[ord(char) - ord('A')] += 1
        return freq_counter


            
        

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-1.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int maxLengthFoundTillNow = 0;

        int left = 0, right = 0;

        if(s.length() == 1)
            return 1;
        else if(s.length() == 0)
            return 0;

        //deal with strings with length >= 2
        Set<Character> set = new HashSet<>();
        while(right < s.length()){
            
            // Remove characters from left till we can fit in the character at index = right
            while(set.contains(s.charAt(right))){
                set.remove(s.charAt(left));
                left++;
            }

            set.add(s.charAt(right));

            int currLen = right - left + 1;
            maxLengthFoundTillNow = Math.max(maxLengthFoundTillNow, currLen);

            right++;
        }

        return maxLengthFoundTillNow;
    }
}

// left = 0, right = 0
// left = 0, right = 1 set = {a}
// left = 0, right = 2 set = {a, b}
// left = 0, right = 3 set = {a, b, c}